qiskit/algorithms/minimum_eigen_solvers/vqe.py

```diff
@@ -40,14 +41,30 @@
 from qiskit.utils.backend_utils import is_aer_provider
 from qiskit.utils.deprecation import deprecate_function
 from qiskit.utils import QuantumInstance, algorithm_globals
-from ..optimizers import Optimizer, SLSQP
+from ..optimizers import Optimizer, SLSQP, OptimizerResult
 from ..variational_algorithm import VariationalAlgorithm, VariationalResult
 from .minimum_eigen_solver import MinimumEigensolver, MinimumEigensolverResult, ListOrDict
 from ..exceptions import AlgorithmError
 
 logger = logging.getLogger(__name__)
 
 
+OBJECTIVE = Callable[[np.ndarray], float]
+GRADIENT = Callable[[np.ndarray], np.ndarray]
+RESULT = Union[scipy.optimize.OptimizeResult, OptimizerResult]
+BOUNDS = List[Tuple[float, float]]
+
+MINIMIZER = Callable[
+    [
+        OBJECTIVE,  # the objective function to minimize (the energy in the case of the VQE)
+        np.ndarray,  # the initial point for the optimization
+        Optional[GRADIENT],  # the gradient of the objective function
+        Optional[BOUNDS],  # parameters bounds for the optimization
+    ],
+    RESULT,  # a result object (either SciPy's or Qiskit's)
+]
+
```

Comment:
Is this type hint actually correct for any SciPy minimiser?  It seems to imply that the first four arguments are positional, but the first four positional arguments of their minimisers are (objective, x0, args, jac).

In an ideal world, I think we'd put this in `__init__.py`, and write some documentation in the docstring about the types. It can have a Sphinx cross-ref (`.. _algorithms_minimum_eigensolvers_minimizer:`, then you reference with `:ref:`algorithms_minimum_eigensolvers_minimizer`), then the various class docstrings could all link to it.  That said, I don't think you need to worry much about it right now.

Reply:
In fact the callable must have the named arguments `fun, x0, jac, bounds` but I didn't quite know how to put that in a type hint using only what's available now 🤔

Reply:
how about using Callable Protocol? [PEP544](https://www.python.org/dev/peps/pep-0544/#callback-protocols)

Reply:
We tried that but it's only supported from Python 3.8 onwards, see the discussion above: https://github.com/Qiskit/qiskit-terra/pull/7191#issuecomment-966467652

Reply:
Protocol is back ported by typing-extensions. https://pypi.org/project/typing-extensions/ so you can use it.
(I'm sorry I overlooked...)

Reply:
Yeah but that would create an additional dependency to a new package, so we didn't add it (yet) 😄